klippy/kinematics/cartesian.py

```diff
@@ -126,7 +126,7 @@ def _activate_carriage(self, carriage):
         toolhead.flush_step_generation()
         dc_rail = self.dual_carriage_rails[carriage]
         dc_axis = self.dual_carriage_axis
-        self.rails[dc_axis].set_trapq(None)
+        self.dual_carriage_rails[not carriage].set_trapq(None)
```

Comment:
I don't understand this change.  Can you elaborate on what it does?

Also, I suspect it should be in a separate commit (that way it if there is a problem it does not complicate a "git revert").

Reply:
Yes. This is actually a work-around of an issue with dual carriage support initially reported for axes smoothing, but also existing for input shaping. The problematic part is [this one](https://github.com/KevinOConnor/klipper/pull/3032/files#diff-7e6a0286c1f3bbf7493501136a9f7f82R58
):
+            s.set_trapq(self.toolhead.get_trapq())

This code resets trapq on all steppers on connection, including deactivated dual carriage stepper. Now the previous version of the code has self.rails[dc_axis] == dc_rail on initial execution of _activate_carriage. So essentially the previous version of the code did nothing useful here on the first call: it would set dc_rail trapq to None and then immediately to toolhead.get_trapq(). So I took advantage of it and made sure that it always resets trapq of a carriage being deactivated, which also fixes the issue in input_shaper.

Another alternative could be to add an API call to a stepper class to return its active trapq - so then input_shaper code can check whether it's None or not. Would that be a better alternative? And I wonder if it will have other side-effects.

I agree that how it is now, it may be a good idea to commit it separately. If sticking with this approach, would it be better to commit it before this pull request?

Reply:
Ah, okay.  I think we could enhance `stepper.set_stepper_kinematics()` to automatically reregister the existing trapq (if valid).  I think input_shaper.py then wouldn't need to call `set_trapq()` at all.

-Kevin

Reply:
I do not know if that would interfere with [force_move](https://github.com/KevinOConnor/klipper/blob/master/klippy/extras/force_move.py#L82). I could probably use a snippet there:
```
prev_trapq = stepper.set_trapq(self.trapq)
prev_sk = stepper.set_stepper_kinematics(self.stepper_kinematics)
...
stepper.set_trapq(prev_trapq)
stepper.set_stepper_kinematics(prev_sk)
```

Still, do you prefer to do that refactoring in a separate pull request? Before this one?

Reply:
Yes - that's how I would fix force_move.  Probably simplest to commit the change to set_stepper_kinematics first.  I can make that change if you like - I believe it to be:
```diff
--- a/klippy/extras/force_move.py
+++ b/klippy/extras/force_move.py
@@ -39,7 +39,6 @@ class ForceMove:
         self.trapq_free_moves = ffi_lib.trapq_free_moves
         self.stepper_kinematics = ffi_main.gc(
             ffi_lib.cartesian_stepper_alloc('x'), ffi_lib.free)
-        ffi_lib.itersolve_set_trapq(self.stepper_kinematics, self.trapq)
         # Register commands
         gcode = self.printer.lookup_object('gcode')
         gcode.register_command('STEPPER_BUZZ', self.cmd_STEPPER_BUZZ,
@@ -80,6 +79,7 @@ class ForceMove:
         toolhead = self.printer.lookup_object('toolhead')
         toolhead.flush_step_generation()
         prev_sk = stepper.set_stepper_kinematics(self.stepper_kinematics)
+        prev_trapq = stepper.set_trapq(self.trapq)
         stepper.set_position((0., 0., 0.))
         axis_r, accel_t, cruise_t, cruise_v = calc_move_time(dist, speed, accel)
         print_time = toolhead.get_last_move_time()
@@ -88,6 +88,7 @@ class ForceMove:
         print_time = print_time + accel_t + cruise_t + accel_t
         stepper.generate_steps(print_time)
         self.trapq_free_moves(self.trapq, print_time + 99999.9)
+        stepper.set_trapq(prev_trapq)
         stepper.set_stepper_kinematics(prev_sk)
         toolhead.note_kinematic_activity(print_time)
         toolhead.dwell(accel_t + cruise_t + accel_t)
diff --git a/klippy/stepper.py b/klippy/stepper.py
index 4249dccb..ad3a6006 100644
--- a/klippy/stepper.py
+++ b/klippy/stepper.py
@@ -124,8 +124,9 @@ class MCU_stepper:
         old_sk = self._stepper_kinematics
         self._stepper_kinematics = sk
         if sk is not None:
-            self._ffi_lib.itersolve_set_stepcompress(
-                sk, self._stepqueue, self._step_dist)
+            self._ffi_lib.itersolve_set_stepcompress(sk, self._stepqueue,
+                                                     self._step_dist)
+            self.set_trapq(self._trapq)
         return old_sk
     def note_homing_end(self, did_trigger=False):
         ret = self._ffi_lib.stepcompress_reset(self._stepqueue, 0)
```

-Kevin

Reply:
Also perhaps extruder.py will need to be updated? But sure, I don't mind either way.

Reply:
Okay - I committed the change (c6b65204).  I don't think extruder.py needs to change.

-Kevin